lib/client/redirect.go

```diff
@@ -0,0 +1,248 @@
+	}
+	query := u.Query()
+	query.Set("secret_key", rd.key.String())
+	u.RawQuery = query.Encode()
+
+	out, err := rd.proxyClient.PostJSON(rd.Context, rd.proxyClient.Endpoint("webapi", rd.Protocol, "login", "console"), SSOLoginConsoleReq{
+		RedirectURL:   u.String(),
+		PublicKey:     rd.PubKey,
+		CertTTL:       rd.TTL,
+		ConnectorID:   rd.ConnectorID,
+		Compatibility: rd.Compatibility,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var re *SSOLoginConsoleResponse
+	err = json.Unmarshal(out.Bytes(), &re)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// notice late binding of the redirect URL here, it is referenced
+	// in the callback handler, but is known only after the request
+	// is sent to the Teleport Proxy, that's why
+	// redirectURL is a SyncString
+	rd.redirectURL.Set(re.RedirectURL)
+	return nil
+}
+
+// Done is called when redirector is closed
+// or parent context is closed
+func (rd *Redirector) Done() <-chan struct{} {
+	return rd.context.Done()
+}
+
+// ClickableURL returns a short clickable redirect URL
+func (rd *Redirector) ClickableURL() string {
+	if rd.server == nil {
+		return "<undefined - server is not started>"
+	}
+	return utils.ClickableURL(rd.server.URL + rd.shortPath)
+}
+
+// ResponseC returns a channel with response
+func (rd *Redirector) ResponseC() <-chan *auth.SSHLoginResponse {
+	return rd.responseC
+}
+
+// ErrorC returns a channel with error
+func (rd *Redirector) ErrorC() <-chan error {
+	return rd.errorC
+}
+
+// callback is used by Teleport proxy to send back credentials
+// issued by Teleport proxy
+func (rd *Redirector) callback(w http.ResponseWriter, r *http.Request) (*auth.SSHLoginResponse, error) {
+	if r.URL.Path != "/callback" {
+		return nil, trace.NotFound("path not found")
+	}
+
+	// Decrypt ciphertext to get login response.
+	plaintext, err := rd.key.Open([]byte(r.URL.Query().Get("response")))
+	if err != nil {
+		return nil, trace.BadParameter("failed to decrypt response: in %v, err: %v", r.URL.String(), err)
+	}
+
+	var re *auth.SSHLoginResponse
+	err = json.Unmarshal(plaintext, &re)
+	if err != nil {
+		return nil, trace.BadParameter("failed to decrypt response: in %v, err: %v", r.URL.String(), err)
+	}
+
+	return re, nil
+}
+
+// Close closes redirector and releases all resources
+func (rd *Redirector) Close() error {
+	rd.cancel()
+	if rd.server != nil {
+		rd.server.Close()
+	}
+	return nil
+}
+
+// wrapCallback is a helper wrapper method that wraps callback HTTP handler
+// and sends a result to the channel and redirect users to error page
+func (rd *Redirector) wrapCallback(fn func(http.ResponseWriter, *http.Request) (*auth.SSHLoginResponse, error)) http.Handler {
+	clone := *rd.proxyURL
+	clone.Path = "/web/msg/error/login_failed"
+	errorURL := clone.String()
+	clone.Path = "/web/msg/info/login_success"
+	successURL := clone.String()
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		response, err := fn(w, r)
+		if err != nil {
```

Comment:
Is this error logged anywhere before redirect?

Reply:
yes, it's sent to the error channel below